Add jit request API and schema.

dandanwang0320 opened the pull request:
### Description:
Add schema of JIT request related APIs.
Most APIs are from AzureUX-ARM/src/providers/Roles/Providers.Feature/Controllers/JitRequestController.cs 
And "updateAccess" api from
ARM/src/providers/Roles/Providers.Feature/Controllers/ApplicationApiController.cs

### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [ ] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [ ] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [ ] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [ ] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [ ] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

dandanwang0320 committed:
add jit request

AutorestCI commented:
# Automation for azure-sdk-for-ruby

AutorestCI commented:
# Automation for azure-sdk-for-js

AutorestCI commented:
# Automation for azure-sdk-for-python

AutorestCI commented:
# Automation for azure-sdk-for-go

azuresdkci commented:
Can one of the admins verify this patch?

adxsdknet commented:
# Automation for azure-sdk-for-net 
A PR has been created for you: 
https://github.com/azure/azure-sdk-for-net/pull/6185 
.NET SDK Commits: 
https://github.com/adxsdknet/azure-sdk-for-net/pull/6185/commits/a5e64370baf2986d86db33db077398b7312b62a6

jhendrixMSFT reviewed (changes requested):
The new swagger isn't referenced in any README.md file so it won't get included in any SDK.  Is this expected?

AutorestCI commented:
# Automation for azure-sdk-for-java

dandanwang0320 committed:
add to readme files.

dandanwang0320 commented:
> The new swagger isn't referenced in any README.md file so it won't get included in any SDK. Is this expected?

jhendrixMSFT committed:
fix swagger path

jhendrixMSFT commented:
Suppressions are added only if there's a linter error that can't be fixed for some legitimate reason, so no need to add anything there at present.

dandanwang0320 committed:
Fix enum and version name

jhendrixMSFT commented on specification/resources/resource-manager/Microsoft.Solutions/stable/2019-03-01/managedapplications.json:
```diff
@@ -0,0 +1,1889 @@
+        "Failed",
+        "Canceled",
+        "Deleted",
+        "Canceled",
```
Duplicate entry needs to be removed.

jhendrixMSFT commented on specification/resources/resource-manager/Microsoft.Solutions/stable/2019-03-01/managedapplications.json:
```diff
@@ -0,0 +1,1889 @@
+                "$ref": "#/definitions/JitRequestDefinitionListResult"
+              }
+            },         
+            "404": {
```
This indicates that a 404 should be treated as a success, is that intentional?  ARM feedback will be interesting here.

dandanwang0320 replied:
I was following the convention of previous versions and other ARM APIs. Is there a general guideline for 404?

jhendrixMSFT replied:
I've not seen this before so it was a surprise to me, doesn't necessarily mean it's incorrect.  The ARM reviewer can provide a definitive answer.

dandanwang0320 replied:
Do you know who's the ARM reviewer? If it's from ARM team, I could talk to them offline.

jhendrixMSFT replied:
Looks like one hasn't been assigned yet.  If this is urgent you can email the ARM on-call person (there are instructions on how to find this info in the ARM API Review Checklist comment at the top.

jhendrixMSFT commented on specification/resources/resource-manager/Microsoft.Solutions/stable/2019-03-01/managedapplications.json:
```diff
@@ -0,0 +1,1889 @@
+      }     
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Solutions/jitRequests": {
+      "get": {
```
Extra "get" should be removed.

dandanwang0320 committed:
Merge branch 'AddJitRequest' of https://github.com/dandanwang0320/azure-rest-api-specs into AddJitRequest

dandanwang0320 committed:
Remove extra "get"

dandanwang0320 committed:
Fix capitalization

jhendrixMSFT commented:
There are model validation failures that need to be fixed, please see the log [here](https://dev.azure.com/azure-sdk/public/_build/results?buildId=28578).

dandanwang0320 commented:
> There are model validation failures that need to be fixed, please see the log [here](https://dev.azure.com/azure-sdk/public/_build/results?buildId=28578).

Actually none of those warnings were add my new API or schema. Or is there anything needs to be added for existing apis in new version?

jhendrixMSFT commented:
They will need to be cleaned up at some point, if you don't want to clean them up now (e.g. tight deadline) we can skip it however can you please open an issue to track fixing them?

dandanwang0320 committed:
Fix previous examples for merge validation.

sanjaiganesh commented:
Requested for description .. updating with old and new apis contents to see the diff easily.

dandanwang0320 committed:
Add changes to another version, so we can see the diff of this change.

dandanwang0320 committed:
Changing back version.

dandanwang0320 commented on specification/resources/resource-manager/Microsoft.Solutions/stable/2018-06-01/managedapplications.json:
```diff
@@ -2,7 +2,7 @@
   "info": {
     "title": "ApplicationClient",
-    "version": "2018-06-01",
+    "version": "2019-03-01",
```
Changing it back.

KrisBash commented on specification/resources/resource-manager/Microsoft.Solutions/stable/2019-03-01/managedapplications.json:
```diff
@@ -0,0 +1,1885 @@
+{
```
In the future, please copy the last reviewed API version into the new version folder as the first commit. This enables an expedited review as we can diff the changes between commits.

KrisBash commented on specification/resources/resource-manager/Microsoft.Solutions/stable/2019-03-01/managedapplications.json:
```diff
@@ -0,0 +1,1885 @@
+            }
+          }
+        },
+        "x-ms-long-running-operation": true
```
examples?

KrisBash commented on specification/resources/resource-manager/Microsoft.Solutions/stable/2019-03-01/managedapplications.json:
```diff
@@ -0,0 +1,1885 @@
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Solutions/applications/{applicationName}/updateAccess": {
+      "post": {
+        "tags": [
```
Is there not a request body?

KrisBash commented on specification/resources/resource-manager/Microsoft.Solutions/stable/2019-03-01/managedapplications.json:
```diff
@@ -0,0 +1,1885 @@
+              "$ref": "#/definitions/JitRequestDefinition"
+            }
+          },
+          "404": {
```
The `default` error is sufficient, there is no need to model specific error codes

KrisBash commented on specification/resources/resource-manager/Microsoft.Solutions/stable/2019-03-01/managedapplications.json:
```diff
@@ -0,0 +1,1885 @@
+      ],
+      "description": "The managed application provider authorization."
+    },
+    "ErrorResponse": {
```
This does not appear to align to RPC error contract, see: https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/common-api-details.md#error-response-content

dandanwang0320 replied:
This actually comes with original schema, not changed by this PR.

KrisBash commented on specification/resources/resource-manager/Microsoft.Solutions/stable/2019-03-01/managedapplications.json:
```diff
@@ -0,0 +1,1885 @@
+          }
+        }
+      },
+      "put": {
```
This is modeled as a tracked resource. Thus, it requires a PATCH api that supports update of "tags", but would ideally support update of all mutable properties

dandanwang0320 replied:
We don't have a patch api for this resource in our code base. Do you mean we should add that to our back end code?

KrisBash replied:
Yes - portal/cli will not be able to update tags on the resource without it

KrisBash commented on specification/resources/resource-manager/Microsoft.Solutions/stable/2019-03-01/managedapplications.json:
```diff
@@ -0,0 +1,1885 @@
+          }
+        ],
+        "responses": {
+          "201": {
```
missing "x-ms-long-running-operation" ?

KrisBash commented on specification/resources/resource-manager/Microsoft.Solutions/stable/2019-03-01/managedapplications.json:
```diff
@@ -0,0 +1,1885 @@
+        ],
+        "operationId": "jitRequests_CreateOrUpdate",
+        "description": "Creates or updates the JIT request.",
+        "parameters": [
```
Where's the request body?

KrisBash reviewed (changes requested):
Please see comments for requested changes

jhendrixMSFT commented:
@dandanwang0320 hello!  Any update here?  Also there's a merge conflict that needs to be resolved.

ravbhatnagar commented:
@dandanwang0320 - please let me know when this is ready for re-review.

dandanwang0320 committed:
PR comments

dandanwang0320 committed:
Merge branch 'master' into AddJitRequest

dandanwang0320 commented:
@KrisBash Hi, would you have time to take another look?

dandanwang0320 committed:
Synced from upstream
Merge branch 'master' into AddJitRequest

# Conflicts:
#	specification/resources/resource-manager/readme.md

dandanwang0320 committed:
Merge branch 'AddJitRequest' of https://github.com/dandanwang0320/azure-rest-api-specs into AddJitRequest

dandanwang0320 committed:
fix build errors.

dandanwang0320 committed:
Fix build errors

dandanwang0320 committed:
Fix build issue.

dandanwang0320 committed:
Fix build errors.

dandanwang0320 committed:
fix model validations

NullMDR commented:
@openapibot sdkautomation rebuild

jhendrixMSFT commented:
@dandanwang0320 hello is this ready for review by ARM?

yungezz commented:
@dandanwang0320 can we close the PR for now since you're not ready for review, and the PR is over several months. Pls create new PR when you're ready for review and merge.

jhendrixMSFT commented:
Closing due to inactivity.
